### PR TITLE
feat: configure nsx global manager mode(active or standby)

### PR DESCRIPTION
**Pull request by slakshmanan2706**
- Added `Set-NsxtGloblaManagerActive` cmdlet to configure the NSX Global Manager to Active mode.
- Added `Set-NsxtGloblaManagerStandby` cmdlet to configure the NSX Global Manager to standby mode.
- Added `Test-NsxVersionCompatibility` cmdlet to check the NSX version compatibility between the active and standby NSX Global Manager.
- Added `Get-NsxtGlobalManagerMode` cmdlet to retrieve the mode of the NSX Global Manager.
- Added `Add-NsxtGlobalManagerMode` cmdlet to configure the NSX Global Manager to Active or Standby mode.

### Type

- [ ] Bugfix
- [x] Enhancement or Feature
- [ ] Code Style or Formatting
- [ ] Documentation
- [ ] Refactoring
- [ ] Chore
- [ ] Other
        Please describe:

### Breaking Changes?

- [ ] Yes, there are breaking changes.
- [x] No, there are no breaking changes.

### Test and Documentation

- [x] Tests have been completed.
- [ ] Documentation has been added or updated.

### CODE REVIEW

File: PowerValidatedSolutions.psm1 [hunk 1]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+               $gmMode= "ACTIVE"
+               $body = @{
+                    display_name = $displayName
+                    mode         = "ACTIVE"
+                } | ConvertTo-Json -Depth 2
+                $uri = "https://$nsxtManager/global-manager/api/v1/global-infra/global-managers/$displayName"
+                $response = Invoke-RestMethod -Uri $uri -Method PATCH -Headers $nsxtHeaders -body $body -SkipCertificateCheck 
+            if ($response) { 
+                Write-Output "Configuring the $nsxtManager NSX Global Manager as ($gmMode) : SUCCESSFUL"
+            }
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32137]: Remove this code, we have not typically added Write-Output to API functions

File: PowerValidatedSolutions.psm1 [hunk 2]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+
+    Try {
+        if ($nsxtHeaders.Authorization) {
+               $gmMode= "ACTIVE"
+               $body = @{
+                    display_name = $displayName
+                    mode         = "ACTIVE"
+                } | ConvertTo-Json -Depth 2
+                $uri = "https://$nsxtManager/global-manager/api/v1/global-infra/global-managers/$displayName"
+                $response = Invoke-RestMethod -Uri $uri -Method PATCH -Headers $nsxtHeaders -body $body -SkipCertificateCheck
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32134]:
```suggestion
                Invoke-RestMethod -Uri $uri -Method PATCH -Headers $nsxtHeaders -body $body -SkipCertificateCheck 
```

File: PowerValidatedSolutions.psm1 [hunk 3]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+Function Set-NsxtGloblaManagerActive {
+    <#
+        .SYNOPSIS
+        Configures the NSX Global Manager to Active mode.
+
+        .DESCRIPTION
+        The Set-NsxtGloblaManagerActive cmdlet configures the NSX Global Manager to Active mode.
+
+        .EXAMPLE
+        Set-NsxtGloblaManagerActive -displayName "SF001"
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32115]:
```suggestion
        Set-NsxtGloblaManagerActive -displayName sfo-m01-nsx-gm01 
```

Comment (GaryJBlake) [reply PowerValidatedSolutions.psm1:32115]: Let's use the PnP values and not the one's Hari defined in his code

File: PowerValidatedSolutions.psm1 [hunk 4]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+    <#
+        .SYNOPSIS
+        Configures the NSX Global Manager to Active mode.
+
+        .DESCRIPTION
+        The Set-NsxtGloblaManagerActive cmdlet configures the NSX Global Manager to Active mode.
+
+        .EXAMPLE
+        Set-NsxtGloblaManagerActive -displayName "SF001" 
+        This example configures the NSX Global Manager to Active mode with the display name SFO01.
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32116]:
```suggestion
        This example sets the NSX Global Manager to Active mode with the display name sfo-m01-nsx-gm01.
```

File: PowerValidatedSolutions.psm1 [hunk 5]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+Function Set-NsxtGloblaManagerStandby {
+    <#
+        .SYNOPSIS
+        Configures the NSX Global Manager to standby mode.
+
+        .DESCRIPTION
+        The Set-NsxtGlobalManagerMode cmdlet configures the NSX Global Manager to standby mode.
+
+        .EXAMPLE
+        Set-NsxtGlobalManagerMode -displayName "LAX01" -standbyServer  "lax-m01-nsx-gm01.lax.rainpole.io" -standbyServerUser admin -standbyServerPass "VMw@re123!VMw@re123!"
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32157]:
```suggestion
        Set-NsxtGlobalManagerMode -displayName lax-m01-nsx-gm01 -standbyServer  "lax-m01-nsx-gm01.lax.rainpole.io" -standbyServerUser admin -standbyServerPass "VMw@re1!VMw@re1!"
```

Comment (GaryJBlake) [reply PowerValidatedSolutions.psm1:32157]: Let's use PnP values and passwords that match all the other functions

File: PowerValidatedSolutions.psm1 [hunk 6]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+                $body | Add-Member -Notepropertyname 'display_name' -Notepropertyvalue $displayName
+                $body | Add-Member -Notepropertyname 'connection_info' -Notepropertyvalue $conn_info
+                $body | Add-Member -Notepropertyname 'mode' -Notepropertyvalue "STANDBY"
+                $body = $body | ConvertTo-Json -Depth 5
+                $uri = "https://$nsxtManager/global-manager/api/v1/global-infra/global-managers/$displayName"
+                $response = Invoke-RestMethod -Uri $uri -Method PUT -Headers $Global:nsxtHeaders -body $body -SkipCertificateCheck 
+            # Check the result of setting the secondary Global Manager as standby
+            if ($response) { 
+                Write-Output "Configuring the $nsxtManager NSX Global Manager as STANDBY : SUCCESSFUL"
+            }
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32201]: Remove as per other comment

File: PowerValidatedSolutions.psm1 [hunk 7]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+                $connection_info | Add-Member -Notepropertyname 'thumbprint' -Notepropertyvalue $gmnodethumbprint
+                $conn_info = @()
+                $conn_info += $connection_info
+                $body = New-Object -TypeName psobject
+                $body | Add-Member -Notepropertyname 'display_name' -Notepropertyvalue $displayName
+                $body | Add-Member -Notepropertyname 'connection_info' -Notepropertyvalue $conn_info
+                $body | Add-Member -Notepropertyname 'mode' -Notepropertyvalue "STANDBY"
+                $body = $body | ConvertTo-Json -Depth 5
+                $uri = "https://$nsxtManager/global-manager/api/v1/global-infra/global-managers/$displayName"
+                $response = Invoke-RestMethod -Uri $uri -Method PUT -Headers $Global:nsxtHeaders -body $body -SkipCertificateCheck
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32197]:
```suggestion
                Invoke-RestMethod -Uri $uri -Method PUT -Headers $Global:nsxtHeaders -body $body -SkipCertificateCheck 
```

File: PowerValidatedSolutions.psm1 [hunk 8]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+Function Test-NsxVersionCompatibility {
+    <#
+        .SYNOPSIS
+        Checks the NSX version compatibility between the active and standby NSX Global Manager.
+
+        .DESCRIPTION
+        The Test-NsxVersionCompatibility cmdlet checks the NSX version compatibility between the active and standby NSX Global Manager.
+
+        .EXAMPLE
+        Test-NsxVersionCompatibility -standbyServer  "lax-m01-nsx-gm01.lax.rainpole.io" -standbyServerUser admin -standbyServerPass "VMw@re123!VMw@re123!"
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32221]:
```suggestion
        Test-NsxVersionCompatibility -standbyServer  "lax-m01-nsx-gm01.lax.rainpole.io" -standbyServerUser admin -standbyServerPass "VMw@re1!VMw@re1!"
```

File: PowerValidatedSolutions.psm1 [hunk 9]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$displayName,  
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$standbyServer,
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$standbyServerUser,
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$standbyServerPass
+    )
+
+    Try {
+        if ($nsxtHeaders.Authorization) {
+                #Custom Object to create json body for Connection Info
+                $gmnodethumbprint = (Get-SHA256Thumbprint -url "https://$standbyServer").replace(":", "")
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32183]:
```suggestion
                $gmNodeThumbprint = (Get-SHA256Thumbprint -url "https://$standbyServer").replace(":", "")
```

File: PowerValidatedSolutions.psm1 [hunk 10]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+                $connection_info = New-Object -TypeName psobject
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32184]:
```suggestion
                $connectionDetails = New-Object -TypeName psobject
```

File: PowerValidatedSolutions.psm1 [hunk 11]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+                $connection_info | Add-Member -Notepropertyname 'fqdn' -Notepropertyvalue $standbyServer
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32185]:
```suggestion
                $connectionDetails | Add-Member -Notepropertyname 'fqdn' -Notepropertyvalue $standbyServer
```

File: PowerValidatedSolutions.psm1 [hunk 12]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+                $connection_info | Add-Member -Notepropertyname 'username' -Notepropertyvalue $user
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32186]:
```suggestion
                $connectionDetails | Add-Member -Notepropertyname 'username' -Notepropertyvalue $user
```

File: PowerValidatedSolutions.psm1 [hunk 13]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+                $connection_info | Add-Member -Notepropertyname 'password' -Notepropertyvalue $pass
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32187]:
```suggestion
                $connectionDetails | Add-Member -Notepropertyname 'password' -Notepropertyvalue $pass
```

File: PowerValidatedSolutions.psm1 [hunk 14]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+                $connection_info | Add-Member -Notepropertyname 'thumbprint' -Notepropertyvalue $gmnodethumbprint
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32188]:
```suggestion
                $connectionDetails | Add-Member -Notepropertyname 'thumbprint' -Notepropertyvalue $gmNodeThumbprint
```

File: PowerValidatedSolutions.psm1 [hunk 15]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+                $conn_info = @()
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32189]:
```suggestion
                $connection = @()
```

File: PowerValidatedSolutions.psm1 [hunk 16]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+                $conn_info += $connection_info
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32190]:
```suggestion
                $connection += $connectionDetails
```

File: PowerValidatedSolutions.psm1 [hunk 17]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+                $body = New-Object -TypeName psobject
+                $body | Add-Member -Notepropertyname 'display_name' -Notepropertyvalue $displayName
+                $body | Add-Member -Notepropertyname 'connection_info' -Notepropertyvalue $conn_info
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32193]:
```suggestion
                $body | Add-Member -Notepropertyname 'connection_info' -Notepropertyvalue $connection
```

File: PowerValidatedSolutions.psm1 [hunk 18]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+        else {
+            Write-Error "Not connected to NSX Local/Global Manager, run Request-NsxtToken and try again"
+        }
+    } Catch {
+        Write-Error $_.Exception.Message
+    }
+}
+Export-ModuleMember -Function Set-NsxtGloblaManagerStandby 
+
+Function Test-NsxVersionCompatibility {
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32212]: This same function can be used to check the compatability for standby and Active. Change inputs to fqdn, username and password

File: PowerValidatedSolutions.psm1 [hunk 19]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+Function Get-NsxtGlobalManagerMode {
+    <#
+        .SYNOPSIS
+        Retrieves the mode of the NSX Global Manager.
+
+        .DESCRIPTION
+        The Get-NsxtGlobalManagerMode cmdlet retrieves the mode of NSX Global Manager.
+
+        .EXAMPLE
+        Get-NsxtGlobalManagerMode -displayName "SF001"
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32270]:
```suggestion
        Get-NsxtGlobalManagerMode -displayName sfo-m01-nsx-gm01
```

File: PowerValidatedSolutions.psm1 [hunk 20]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+        .SYNOPSIS
+        Configures the NSX Global Manager to active or standby mode.
+
+        .DESCRIPTION
+        The Add-NsxtGlobalManagerMode cmdlet configures the NSX Global Manager to active or standby mode.
+        - Validates that network connectivity and authentication is possible to NSX Global Manager
+        - Configures the NSX Global Manager to active or standby mode
+
+        .EXAMPLE
+        Add-NsxtGlobalManagerMode -server sfo-m01-nsx-gm01.sfo.rainpole.io -user admin -pass "VMw@re123!VMw@re123!" -gmMode STANDBY -displayName LAX01 -standbyServer sfo-m01-nsx-gm01c.sfo.rainpole.io -standbyServerUser admin1 -standbyServerPass "VMw@re123!VMw@re123!"
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32307]:
```suggestion
        Add-NsxtGlobalManagerMode -server sfo-m01-nsx-gm01.sfo.rainpole.io -user admin -pass "VMw@re1!VMw@re1!" -gmMode STANDBY -displayName lax-m01-nsx-gm01 -standbyServer lax-m01-nsx-gm01c.lax.rainpole.io -standbyServerUser admin1 -standbyServerPass "VMw@re1!VMw@re1!"
```

File: PowerValidatedSolutions.psm1 [hunk 21]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+        This example configures the virtual IP address of the NSX Global Manager cluster.
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32308]: Example description needs correcting

File: PowerValidatedSolutions.psm1 [hunk 22]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+    <#
+        .SYNOPSIS
+        Configures the NSX Global Manager to active or standby mode.
+
+        .DESCRIPTION
+        The Add-NsxtGlobalManagerMode cmdlet configures the NSX Global Manager to active or standby mode.
+        - Validates that network connectivity and authentication is possible to NSX Global Manager
+        - Configures the NSX Global Manager to active or standby mode
+
+        .EXAMPLE
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32306]: We should add an example for setting the ACTIVE node too

File: PowerValidatedSolutions.psm1 [hunk 23]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+        .PARAMETER server
+        The fully qualified domain name of the NSX Global Manager.
+
+        .PARAMETER user
+        The username to authenticate to the NSX Global Manager.
+
+        .PARAMETER pass
+        The password to authenticate to the NSX Global Manager.
+
+        .PARAMETER gmMode
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32319]: Just use mode no need for gm at the front

File: PowerValidatedSolutions.psm1 [hunk 24]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$pass,
+        [Parameter (Mandatory = $true)] [ValidateSet("ACTIVE", "STANDBY")] [String]$gmMode,
+        [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$displayName,
+        [Parameter (Mandatory = $false)] [ValidateNotNullOrEmpty()] [String]$standbyServer,
+        [Parameter (Mandatory = $false)] [ValidateNotNullOrEmpty()] [String]$standbyServerUser,
+        [Parameter (Mandatory = $false)] [ValidateNotNullOrEmpty()] [String]$standbyServerPass
+    )
+
+    Try {
+        if (Test-NSXTConnection -server $server) {
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32347]:
```suggestion
        if (Test-NsxtConnection -server $server) {
```

File: PowerValidatedSolutions.psm1 [hunk 25]
```diff
@@ -32103,6 +32103,277 @@ Function Undo-NsxtRemoteTunnelEndpoint {
...
+            if (Test-NSXTAuthentication -server $server -user $user -pass $pass) {
```
Comment (GaryJBlake) [inline PowerValidatedSolutions.psm1:32348]:
```suggestion
            if (Test-NsxtAuthentication -server $server -user $user -pass $pass) {
```

**Review by GaryJBlake — CHANGES_REQUESTED** [review]

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
- [ ] Update the module manifest build.
- [ ] Update the module manifiest with required function exports.
- [ ] Add the documentation for each exported function.
- [ ] Update the `CHANGELOG.md` if targets for v2.12.0.

**Review by GaryJBlake — APPROVED** [review]
LGTM!

**Pull request merged**